File: apps/backend/app/schemas/proposal_checklist.py
"""Schemas for a proposal's EOT-documentation checklist.

The checklist answers (all items with their status/remarks) are read and saved as
one document via GET/PUT — matching how the UI edits it (locally, then one save)
and how the sibling costing sheet works. The item TEXT is canonical and owned by
the service; only the editable state is sent back on save.

File attachments are the exception: they are immediate single-item operations
(attach / detach / add row), because a file upload can't sensibly sit in an
unsaved local buffer. Every one of them returns the whole refreshed checklist.
"""
from typing import List, Literal, Optional

from pydantic import BaseModel

# Yes = provided, No = not provided, na = not applicable, "" = not yet answered.
ChecklistStatus = Literal["yes", "no", "na", ""]


class ChecklistDocOut(BaseModel):
    """A file attached to a checklist item. It is an ordinary project document —
    it also appears in the proposal's Documents tab and can be analysed there."""

    id: str
    name: str = ""
    type: str = ""
    uploadedAt: str = ""
    uploadedBy: str = ""
    sizeKB: int = 0


class ChecklistItemIn(BaseModel):
    """Editable state for one checklist row, keyed by its item number."""

    no: int
    status: ChecklistStatus = ""
    remarks: str = ""
    # Only honoured for user-added rows (28+); canonical item text is server-owned.
    item: str = ""


class ChecklistItemOut(BaseModel):
    no: int
    # The canonical (or user-set) item text, merged in from the service on read.
    item: str = ""
    status: ChecklistStatus = ""
    remarks: str = ""
    documents: List[ChecklistDocOut] = []


class ChecklistOut(BaseModel):
    projectId: str
    items: List[ChecklistItemOut] = []
    updatedAt: Optional[str] = None


class ChecklistIn(BaseModel):
    """Replace the whole checklist state for a proposal in one call."""

    items: List[ChecklistItemIn] = []


class AttachDocIn(BaseModel):
    """Link an already-uploaded project document to a checklist item."""

    documentId: str
    # Names the row. Only honoured for the additional-document rows (27+).
    item: str = ""


class NewItemIn(BaseModel):
    """Add a further "additional document" row (28, 29, …)."""

    item: str = ""


class NewItemOut(BaseModel):
    """The refreshed checklist plus the number of the row just created, so the
    caller can attach the file it just uploaded to it."""

    no: int
    checklist: ChecklistOut

File: apps/web/src/api/proposalChecklist.ts
import { api } from "./client";

/** Answer for a checklist row. "" = not yet answered. */
export type ChecklistStatus = "yes" | "no" | "na" | "";

/** A file attached to a checklist item. It is an ordinary project document, so it
 *  also shows up in the proposal's Documents tab. */
export interface ChecklistDocument {
  id: string;
  name: string;
  type: string;
  uploadedAt: string;
  uploadedBy: string;
  sizeKB: number;
}

/** One row of the EOT-documentation checklist: the item text plus the editable
 *  state (status / remarks) and the documents attached to it. */
export interface ChecklistItem {
  no: number;
  item: string;
  status: ChecklistStatus;
  remarks: string;
  documents: ChecklistDocument[];
}

export interface ProposalChecklist {
  projectId: string;
  items: ChecklistItem[];
  updatedAt: string | null;
}

/** State-only payload sent on save. Canonical item text and the attached document
 *  ids are owned by the server; `item` is only honoured for user-added rows. */
export interface ChecklistItemIn {
  no: number;
  status: ChecklistStatus;
  remarks: string;
  item?: string;
}

/** The "Any additional documents if deemed necessary." row — the first extra file
 *  lands here; every further one gets its own row (28, 29, …). */
export const ADDITIONAL_ITEM_NO = 27;
/** Items 1..26 are the fixed standard documents. */
export const FIXED_ITEM_COUNT = 26;

export async function getChecklistApi(projectId: string): Promise<ProposalChecklist> {
  const { data } = await api.get<ProposalChecklist>(`/proposal-checklist/project/${projectId}`);
  return data;
}

export async function saveChecklistApi(
  projectId: string,
  items: ChecklistItemIn[],
): Promise<ProposalChecklist> {
  const { data } = await api.put<ProposalChecklist>(`/proposal-checklist/project/${projectId}`, { items });
  return data;
}

/** Add a further "additional document" row; returns its number and the checklist. */
export async function addChecklistItemApi(
  projectId: string,
  item = "",
): Promise<{ no: number; checklist: ProposalChecklist }> {
  const { data } = await api.post(`/proposal-checklist/project/${projectId}/items`, { item });
  return data;
}

/** Delete a user-added row (its files stay in the Documents tab). */
export async function removeChecklistItemApi(projectId: string, no: number): Promise<ProposalChecklist> {
  const { data } = await api.delete<ProposalChecklist>(`/proposal-checklist/project/${projectId}/items/${no}`);
  return data;
}

/** Link an already-uploaded project document to a checklist item. `item` names the
 *  row and is only honoured for the additional-document rows (27+). */
export async function attachChecklistDocApi(
  projectId: string,
  no: number,
  documentId: string,
  item = "",
): Promise<ProposalChecklist> {
  const { data } = await api.post<ProposalChecklist>(
    `/proposal-checklist/project/${projectId}/items/${no}/documents`,
    { documentId, item },
  );
  return data;
}

/** Unlink a file from an item — the document itself is not deleted. */
export async function detachChecklistDocApi(
  projectId: string,
  no: number,
  documentId: string,
): Promise<ProposalChecklist> {
  const { data } = await api.delete<ProposalChecklist>(
    `/proposal-checklist/project/${projectId}/items/${no}/documents/${documentId}`,
  );
  return data;
}
